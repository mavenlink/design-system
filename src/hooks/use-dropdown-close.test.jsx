import React, {
  useRef,
  useState,
} from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import useDropdownClose from './use-dropdown-close.js';

describe('useDropdownClose', () => {
  function TestComponent() {
    const ref = useRef();
    const [opened, setOpened] = useState(false);

    function onClose() {
      setOpened(false);
    }

    function onFocus() {
      setOpened(true);
    }

    useDropdownClose(ref, opened, onClose);

    return (
      <div onFocus={onFocus} ref={ref} tabIndex="0">{opened ? 'opened' : 'closed'}</div>
    );
  }

  it('can be used in a component', () => {
    render(<TestComponent />);
    expect(document.body).toMatchSnapshot();
  });

  it('stays opens', () => {
    render(<TestComponent />);
    userEvent.tab();
    expect(screen.getByText('opened')).toBeInTheDocument();
    userEvent.click(screen.getByText('opened'));
    expect(screen.getByText('opened')).toBeInTheDocument();
  });

  it('closes on blur', () => {
    render((
      <React.Fragment>
        <TestComponent />
        <input />
      </React.Fragment>
    ));
    userEvent.tab();
    expect(screen.getByText('opened')).toBeInTheDocument();
    userEvent.tab();
    expect(screen.getByText('closed')).toBeInTheDocument();
  });

  it('closes on click away', () => {
    render(<TestComponent />);
    userEvent.tab();
    expect(screen.getByText('opened')).toBeInTheDocument();
    userEvent.click(document.body);
    expect(screen.getByText('closed')).toBeInTheDocument();
  });
});
