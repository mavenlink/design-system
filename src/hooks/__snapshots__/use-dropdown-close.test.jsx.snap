// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDropdownClose can be used in a component 1`] = `
<body>
  <div>
    <div
      tabindex="0"
    >
      closed
    </div>
  </div>
</body>
`;
